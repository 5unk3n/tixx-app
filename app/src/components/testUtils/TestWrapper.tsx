import { NavigationContainer } from '@react-navigation/native'
import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import React, { PropsWithChildren } from 'react'
import { PaperProvider } from 'react-native-paper'

import { theme } from '@/theme/defaultTheme'

const queryClient = new QueryClient({
	defaultOptions: {
		queries: {
			retry: false,
			staleTime: Infinity
		}
	}
})

export const TestWrapper = ({ children }: PropsWithChildren) => {
	return (
		<PaperProvider theme={theme.dark}>
			<QueryClientProvider client={queryClient}>
				<NavigationContainer>{children}</NavigationContainer>
			</QueryClientProvider>
		</PaperProvider>
	)
}
