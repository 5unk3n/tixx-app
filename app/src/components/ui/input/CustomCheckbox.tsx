import React from 'react'
import { View, ViewProps } from 'react-native'
import { TouchableRipple } from 'react-native-paper'

import { useCustomTheme } from '@/hooks/useCustomTheme'

import CustomIcon from '../display/CustomIcon'
import { CustomText } from '../display/CustomText'

interface CustomCheckboxProps extends ViewProps {
	checked: boolean
	label: string
	onChange: (checked: boolean) => void
	labelSize?: 'md' | 'lg'
	iconType?: 'square' | 'minimal'
	iconPosition?: 'left' | 'right'
}

export default function CustomCheckbox({
	checked,
	label,
	onChange,
	labelSize = 'md',
	iconType = 'minimal',
	iconPosition = 'left',
	...props
}: CustomCheckboxProps) {
	const { colors } = useCustomTheme()
	const iconPrefix =
		iconType === 'square' ? 'checkboxSquare' : 'checkboxMinimal'
	const iconSuffix = checked ? 'Checked' : 'Unchecked'

	return (
		<TouchableRipple onPress={() => onChange(!checked)} className="flex-1">
			<View
				className={`${iconPosition === 'left' ? 'flex-row' : 'flex-row-reverse justify-between'} items-center`}
				{...props}
			>
				<CustomIcon
					name={`${iconPrefix}${iconSuffix}`}
					color={colors.primary}
					className={`${iconPosition === 'left' ? `mr-2` : 'ml-auto'}`}
				/>
				<CustomText
					variant={labelSize === 'lg' ? 'headline1Semibold' : 'body1Medium'}
					className="text-grayscale-400 flex-1"
				>
					{label}
				</CustomText>
			</View>
		</TouchableRipple>
	)
}
