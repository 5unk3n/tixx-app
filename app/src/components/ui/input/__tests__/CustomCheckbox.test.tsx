import { describe, it, expect, jest } from '@jest/globals'
import { render, screen, fireEvent } from '@testing-library/react-native'
import React from 'react'

import { TestWrapper } from '../../../testUtils/TestWrapper'
import CustomCheckbox from '../CustomCheckbox'

describe('CustomCheckbox', () => {
	it('should toggle checked state and render label', () => {
		const onChangeMock = jest.fn()
		render(
			<CustomCheckbox checked={false} label="동의" onChange={onChangeMock} />,
			{ wrapper: TestWrapper }
		)

		// Pressing the label should trigger onChange with the new value
		fireEvent.press(screen.getByText('동의'))
		expect(onChangeMock).toHaveBeenCalledWith(true)
	})

	describe('should reflect icon position and type', () => {
		// Test cases for different props combinations
		const testCases = [
			{ iconPosition: 'left', iconType: 'minimal', checked: true },
			{ iconPosition: 'right', iconType: 'minimal', checked: false },
			{ iconPosition: 'left', iconType: 'square', checked: true },
			{ iconPosition: 'right', iconType: 'square', checked: false }
		] as const

		testCases.forEach((props) => {
			it(`renders correctly for ${JSON.stringify(props)}`, () => {
				const { toJSON } = render(
					<CustomCheckbox label="옵션" onChange={jest.fn()} {...props} />,
					{ wrapper: TestWrapper }
				)
				// Compare with snapshot to verify structural changes (icon order, icon name)
				expect(toJSON()).toMatchSnapshot()
			})
		})
	})
})
