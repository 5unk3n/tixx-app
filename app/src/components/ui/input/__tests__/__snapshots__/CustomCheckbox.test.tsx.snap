// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomCheckbox should reflect icon position and type renders correctly for {"iconPosition":"left","iconType":"minimal","checked":true} 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      {
        "flex": 1,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          false,
          [
            {
              "flexBasis": "0%",
              "flexGrow": 1,
              "flexShrink": 1,
            },
          ],
        ]
      }
    >
      <View
        style={
          [
            {
              "flexDirection": "row",
            },
            {
              "alignItems": "center",
            },
          ]
        }
      >
        <test-file-stub
          color="#F2F862"
          height={26}
          pointerEvents="none"
          style={
            [
              {
                "marginRight": 8,
              },
            ]
          }
          width={26}
        />
        <Text
          style={
            [
              {
                "textAlign": "left",
              },
              {
                "color": "#FFFFFF",
                "writingDirection": "ltr",
              },
              [
                {
                  "fontFamily": "Pretendard-Medium",
                  "fontSize": 16,
                  "fontWeight": "500",
                  "letterSpacing": -0.02,
                  "lineHeight": 18,
                },
                [
                  {
                    "color": "#A0A2A2",
                  },
                  {
                    "flexBasis": "0%",
                    "flexGrow": 1,
                    "flexShrink": 1,
                  },
                ],
              ],
            ]
          }
        >
          옵션
        </Text>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;

exports[`CustomCheckbox should reflect icon position and type renders correctly for {"iconPosition":"left","iconType":"square","checked":true} 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      {
        "flex": 1,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          false,
          [
            {
              "flexBasis": "0%",
              "flexGrow": 1,
              "flexShrink": 1,
            },
          ],
        ]
      }
    >
      <View
        style={
          [
            {
              "flexDirection": "row",
            },
            {
              "alignItems": "center",
            },
          ]
        }
      >
        <test-file-stub
          color="#F2F862"
          height={26}
          pointerEvents="none"
          style={
            [
              {
                "marginRight": 8,
              },
            ]
          }
          width={26}
        />
        <Text
          style={
            [
              {
                "textAlign": "left",
              },
              {
                "color": "#FFFFFF",
                "writingDirection": "ltr",
              },
              [
                {
                  "fontFamily": "Pretendard-Medium",
                  "fontSize": 16,
                  "fontWeight": "500",
                  "letterSpacing": -0.02,
                  "lineHeight": 18,
                },
                [
                  {
                    "color": "#A0A2A2",
                  },
                  {
                    "flexBasis": "0%",
                    "flexGrow": 1,
                    "flexShrink": 1,
                  },
                ],
              ],
            ]
          }
        >
          옵션
        </Text>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;

exports[`CustomCheckbox should reflect icon position and type renders correctly for {"iconPosition":"right","iconType":"minimal","checked":false} 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      {
        "flex": 1,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          false,
          [
            {
              "flexBasis": "0%",
              "flexGrow": 1,
              "flexShrink": 1,
            },
          ],
        ]
      }
    >
      <View
        style={
          [
            {
              "flexDirection": "row-reverse",
            },
            {
              "justifyContent": "space-between",
            },
            {
              "alignItems": "center",
            },
          ]
        }
      >
        <test-file-stub
          color="#F2F862"
          height={26}
          pointerEvents="none"
          style={
            [
              {
                "marginLeft": "auto",
              },
            ]
          }
          width={26}
        />
        <Text
          style={
            [
              {
                "textAlign": "left",
              },
              {
                "color": "#FFFFFF",
                "writingDirection": "ltr",
              },
              [
                {
                  "fontFamily": "Pretendard-Medium",
                  "fontSize": 16,
                  "fontWeight": "500",
                  "letterSpacing": -0.02,
                  "lineHeight": 18,
                },
                [
                  {
                    "color": "#A0A2A2",
                  },
                  {
                    "flexBasis": "0%",
                    "flexGrow": 1,
                    "flexShrink": 1,
                  },
                ],
              ],
            ]
          }
        >
          옵션
        </Text>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;

exports[`CustomCheckbox should reflect icon position and type renders correctly for {"iconPosition":"right","iconType":"square","checked":false} 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      {
        "flex": 1,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          false,
          [
            {
              "flexBasis": "0%",
              "flexGrow": 1,
              "flexShrink": 1,
            },
          ],
        ]
      }
    >
      <View
        style={
          [
            {
              "flexDirection": "row-reverse",
            },
            {
              "justifyContent": "space-between",
            },
            {
              "alignItems": "center",
            },
          ]
        }
      >
        <test-file-stub
          color="#F2F862"
          height={26}
          pointerEvents="none"
          style={
            [
              {
                "marginLeft": "auto",
              },
            ]
          }
          width={26}
        />
        <Text
          style={
            [
              {
                "textAlign": "left",
              },
              {
                "color": "#FFFFFF",
                "writingDirection": "ltr",
              },
              [
                {
                  "fontFamily": "Pretendard-Medium",
                  "fontSize": 16,
                  "fontWeight": "500",
                  "letterSpacing": -0.02,
                  "lineHeight": 18,
                },
                [
                  {
                    "color": "#A0A2A2",
                  },
                  {
                    "flexBasis": "0%",
                    "flexGrow": 1,
                    "flexShrink": 1,
                  },
                ],
              ],
            ]
          }
        >
          옵션
        </Text>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
